feat(reporter): implement plain text issue formatter

ostridm opened the pull request:
closes #97

ostridm committed:
feat(reporter) add plain text issue formatter

closes #97

ostridm committed:
feat(reporter) fix CI test failure on windows-2019

closes #97

ostridm committed:
feat(reporter): add plain text Formatter to IoC

closes #97

derevnjuk commented on src/SecTester.Reporter/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,13 @@
+
+public static class ServiceCollectionExtensions
+{
+  public static IServiceCollection AddSecTesterReporter(this IServiceCollection collection)
```
You don't need to provide a method to register a default formatter. This is runner's responsibility.

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -0,0 +1,98 @@
+
+namespace SecTester.Reporter;
+
+internal class DefaultFormatter : Formatter
```
It is not clear, why it is internal

derevnjuk commented on src/SecTester.Reporter/SecTester.Reporter.csproj:
```diff
@@ -4,6 +4,12 @@
+    <ProjectReference Include="..\SecTester.Scan\SecTester.Scan.csproj" />
+  </ItemGroup>
+
+  <ItemGroup>
+    <AssemblyAttribute Include="System.Runtime.CompilerServices.InternalsVisibleTo">
+      <_Parameter1>$(MSBuildProjectName).Tests</_Parameter1>
+    </AssemblyAttribute>
   </ItemGroup>
```
ditto

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -0,0 +1,98 @@
+{3}
+Details:
+{4}";
+  private const string TemplateExtraDetails = "\nExtra Details:\n{5}";
+  private const string TemplateReferences = "\nReferences:\n{6}";
```
The template cannot contain any trailing/leading new lines

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -0,0 +1,98 @@
+  private const string TemplateExtraDetails = "\nExtra Details:\n{5}";
+  private const string TemplateReferences = "\nReferences:\n{6}";
+
+  private static readonly IEnumerable<Comment> EmptyComments = new List<Comment>();
+  private static readonly IEnumerable<string> EmptyResources = new List<string>();
```
So why does it need? It seems you forgot what we discussed recently about redundant optimization in place of readability and cleanness (e.g. `issue.Comments ?? Array.Empty<Comment>()`)

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -0,0 +1,98 @@
+
+  private static string CombineList(IEnumerable<string> list, string? sep = default)
+  {
+    return string.Join(sep ?? NewLine.ToString(), list);
```
`NewLine` is const, why don't specify it as a value by default in an argument?

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -0,0 +1,98 @@
+  private static string FormatList<T>(IEnumerable<T> list, Func<T, string>? map = default)
+    where T : class
+  {
+    var items = list.Select(
```
Why don't make the `map` optional, like this:

```csharp
var formatItem = map ?? (x => x.ToString());

```

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -0,0 +1,98 @@
+
+  private static string FormatExtraInfo(Comment comment)
+  {
+    var footer = comment.Links is not null && comment.Links.Any() ? CombineList(comment.Links.Prepend("Links:")) : String.Empty;
```
Use `""` instead of String.Empty

```suggestion
    var footer = comment.Links is not null && comment.Links.Any() ? CombineList(comment.Links.Prepend("Links:")) : "";
```

derevnjuk commented on test/SecTester.Reporter.Tests/DefaultFormatterTests.cs:
Why don't inline a tab in the test data?

derevnjuk commented on test/SecTester.Reporter.Tests/DefaultFormatterTests.cs:
```diff
@@ -0,0 +1,76 @@
+using System.Xml;
```
global using

ostridm committed:
feat(reporter) add plain text issue formatter

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046843196
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046843860
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046844123
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046854052
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046856125
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046865904
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046854052
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046872417
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046879039
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046878764
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046847904

closes #97

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -52,12 +49,20 @@ private static string GenerateTemplate(bool extraInfo, bool references)
     if (references)
     {
-      stringBuilder.Append(TemplateReferences);
+      stringBuilder
+        .Append(NewLine)
+        .Append(TemplateReferences)
+        .Append(NewLine)
+        .Append("{6}");
```
You just need to remove trailing or leading, not in the middle https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1046847904

derevnjuk commented on test/SecTester.Reporter.Tests/DefaultFormatterTests.cs:
```suggestion
          new ("CommentHeadline", new List<string> { "https://example.com/comment/1", "https://example.com/comment/2" },
```

derevnjuk commented on test/SecTester.Reporter.Tests/DefaultFormatterTests.cs:
```suggestion
  public static readonly IEnumerable<object[]> FormatInput = new List<object[]>
```

derevnjuk commented on test/SecTester.Reporter.Tests/DefaultFormatterTests.cs:
Why did you split it by lines?

ostridm replied:
Just because `@` modifier skips translation of `\t`

ostridm committed:
feat(reporter): fix windows auto EOL conversion for verbatim strings and get rid of `StringBuilder`

closes #97

ostridm committed:
feat(reporter): fix windows auto EOL conversion for verbatim strings

closes #97

derevnjuk commented on src/SecTester.Reporter/DefaultFormatter.cs:
```diff
@@ -0,0 +1,101 @@
+      templates = templates.Append(TemplateReferences);
+    }
+
+    templates = templates.Select(x => x.Replace("\r\n", NewLine.ToString()));
```
Please add ADHOC comment here enumerating all previous solutions

derevnjuk commented on test/SecTester.Reporter.Tests/DefaultFormatterTests.cs:
ditto

ostridm committed:
feat(reporter) add plain text issue formatter

adjust according to code review
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1047415609
https://github.com/NeuraLegion/sectester-net/pull/112#discussion_r1047415753

closes #97

derevnjuk commented:
@ostridm please resolve conflicts rebasing onto master
